metadata/images_test.go

```diff
@@ -0,0 +1,503 @@
+		},
+	} {
+		t.Run(testcase.name, func(t *testing.T) {
+			testcase.original.Name = testcase.name
+			if testcase.input.Name == "" {
+				testcase.input.Name = testcase.name
+			}
+			testcase.expected.Name = testcase.name
+
+			if testcase.original.Target.Digest == "" {
+				testcase.original.Target.Digest = digest.FromString(testcase.name)
+				testcase.input.Target.Digest = testcase.original.Target.Digest
+				testcase.expected.Target.Digest = testcase.original.Target.Digest
+			}
+
+			done := errors.New("test complete")
+			if err := db.Update(func(tx *bolt.Tx) error {
+				var (
+					store = NewImageStore(tx)
+					now   = time.Now()
+				)
+
+				created, err := store.Create(ctx, testcase.original)
+				if errors.Cause(err) != testcase.createerr {
+					if testcase.createerr == nil {
+						t.Fatalf("unexpected error: %v", err)
+					} else {
+						t.Fatalf("cause of %v (cause: %v) != %v", err, errors.Cause(err), testcase.createerr)
+					}
+				} else if testcase.createerr != nil {
+					return done
+				}
+
+				checkImageTimestamps(t, &created, now, true)
+
+				testcase.original.CreatedAt = created.CreatedAt
+				testcase.expected.CreatedAt = created.CreatedAt
+				testcase.original.UpdatedAt = created.UpdatedAt
+				testcase.expected.UpdatedAt = created.UpdatedAt
+
+				checkImagesEqual(t, &created, &testcase.original, "unexpected image on creation")
+				return nil
+			}); err != nil {
+				if err == done {
+					return
+				}
+				t.Fatal(err)
+			}
+
+			if err := db.Update(func(tx *bolt.Tx) error {
+				now := time.Now()
+				store := NewImageStore(tx)
+				updated, err := store.Update(ctx, testcase.input, testcase.fieldpaths...)
+				if errors.Cause(err) != testcase.cause {
+					if testcase.cause == nil {
+						t.Fatalf("unexpected error: %v", err)
+					} else {
+						t.Fatalf("cause of %v (cause: %v) != %v", err, errors.Cause(err), testcase.cause)
+					}
+				} else if testcase.cause != nil {
+					return done
+				}
+
+				checkImageTimestamps(t, &updated, now, false)
+				testcase.expected.UpdatedAt = updated.UpdatedAt
+				checkImagesEqual(t, &updated, &testcase.expected, "updated failed to get expected result")
+				return nil
+			}); err != nil {
+				if err == done {
+					return
+				}
+				t.Fatal(err)
+			}
+
+			if err := db.View(func(tx *bolt.Tx) error {
+				store := NewImageStore(tx)
+				result, err := store.Get(ctx, testcase.original.Name)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				checkImagesEqual(t, &result, &testcase.expected, "get after failed to get expected result")
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+
+		})
+	}
+}
+
+func imageBase() images.Image {
+	return images.Image{
+		Labels: map[string]string{
+			"foo": "bar",
+			"baz": "boo",
+		},
+		Target: ocispec.Descriptor{
+			Size:      10,
+			MediaType: "application/vnd.oci.blab",
+		},
+	}
+}
+
+func checkImageTimestamps(t *testing.T, im *images.Image, now time.Time, oncreate bool) {
+	t.Helper()
```

Comment:
when I ran `make test` after pulling this I got a `t.Helper undefined`: should that happen?

Reply:
@jessvalarezo it was just added in go 1.9